opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -37,12 +37,29 @@ namespace Opm {
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::
-    beginReportStep(const int timeStepIdx)
+    beginReportStep(const int timeStepIdx,const double simulationTime, bool enableWellTests)
     {
         const Grid& grid = ebosSimulator_.vanguard().grid();
         const auto& defunct_well_names = ebosSimulator_.vanguard().defunctWellNames();
         const auto& eclState = ebosSimulator_.vanguard().eclState();
         wells_ecl_ = schedule().getWells(timeStepIdx);
+        const auto& wtest_config = schedule().wtestConfig(timeStepIdx);
+
+        // Check if the well is scheduled for a test
+        if (wtest_config.size() > 0 && enableWellTests) {
+            for (auto& well : wtest_config.getWells()) {
+                wellTestState_.addClosedWell(well.name, WellTestConfig::Reason::ECONOMIC, 0);
```

Comment:
This looks wrong; the call to `WellTestState::addClosedWell()` should be *when you actually close a well*.

Reply:
From the simulator side it is currently the dynamic_list_econ_limited_ that keep track of when a well is closed or not. What I use wellTestState_ for is to check whether the wells should be tested again or not. I add all wells specified in WTEST. Then I call the update method to get the wells ready for testing. Then I remove these wells from the dynamic_list_econ_limited list. Solve the well model. Close the well again if they still violate the *ECON constrains and move on to the real time step.

Reply:
Well - that was some workflow ...